File: packages/mirascript/src/vm/operations.ts
import { VmError } from '../helpers/error.js';
import { hasOwnEnumerable, isNaN, isSafeInteger, keys, create } from '../helpers/utils.js';
import { toNumber, toString, toBoolean, toFormat } from '../helpers/convert/index.js';
import { display } from '../helpers/serialize.js';
import { isVmPrimitive, isVmArray, isVmRecord, isVmFunction, isVmExtern, isVmWrapper } from '../helpers/types.js';
import type { TypeName, VmAny, VmImmutable, VmRecord, VmValue, VmArray, VmConst } from './types/index.js';

const { abs, min, trunc, ceil } = Math;
const { slice, at } = Array.prototype;

const isSame = (a: VmValue, b: VmValue): boolean => {
    // Check for NaN
    if (typeof a == 'number' && typeof b == 'number') return a === b || (isNaN(a) && isNaN(b));
    // Check all primitive types, and fast path for reference equality
    if (a === b) return true;
    // Any primitives and functions arrive here are not equal
    if (a == null || typeof a != 'object' || b == null || typeof b != 'object') return false;
    // Handle wrapper values
    if (isVmWrapper(a)) return a.same(b);
    if (isVmWrapper(b)) return b.same(a);
    // Handle array values
    if (isVmArray(a) && isVmArray(b)) {
        const len = a.length;
        if (len !== b.length) {
            return false;
        }
        // Compare array items
        for (let i = 0; i < len; i++) {
            if (!isSame(a[i] ?? null, b[i] ?? null)) {
                return false;
            }
        }
        return true;
    }
    // Handle record values
    if (!isVmArray(a) && !isVmArray(b)) {
        // Compare record fields
        const aKeys = keys(a);
        const bKeys = keys(b);
        if (aKeys.length !== bKeys.length) {
            return false;
        }
        for (const key of aKeys) {
            if (!hasOwnEnumerable(b, key)) {
                return false;
            }
            /* c8 ignore next 2 */
            const av = a[key] ?? null;
            const bv = b[key] ?? null;
            if (!isSame(av, bv)) {
                return false;
            }
        }
        return true;
    }
    return false;
};

const overloadNumberString = (a: VmAny, b: VmAny): boolean => {
    if (typeof a == 'number' || typeof b == 'number') return true;
    if (typeof a == 'string' || typeof b == 'string') return false;
    return true;
};

// Math operations
export const $Add = (a: VmAny, b: VmAny): number => {
    return $ToNumber(a) + $ToNumber(b);
};
export const $Sub = (a: VmAny, b: VmAny): number => {
    return $ToNumber(a) - $ToNumber(b);
};
export const $Mul = (a: VmAny, b: VmAny): number => {
    return $ToNumber(a) * $ToNumber(b);
};
export const $Div = (a: VmAny, b: VmAny): number => {
    return $ToNumber(a) / $ToNumber(b);
};
export const $Mod = (a: VmAny, b: VmAny): number => {
    return $ToNumber(a) % $ToNumber(b);
};
export const $Pow = (a: VmAny, b: VmAny): number => {
    return $ToNumber(a) ** $ToNumber(b);
};
// Logical operations without short-circuiting
export const $And = (a: VmAny, b: VmAny): boolean => {
    return $ToBoolean(a) && $ToBoolean(b);
};
export const $Or = (a: VmAny, b: VmAny): boolean => {
    return $ToBoolean(a) || $ToBoolean(b);
};
// Comparison operations
export const $Gt = (a: VmAny, b: VmAny): boolean => {
    if (overloadNumberString(a, b)) {
        return $ToNumber(a) > $ToNumber(b);
    } else {
        return $ToString(a) > $ToString(b);
    }
};
export const $Gte = (a: VmAny, b: VmAny): boolean => {
    if (overloadNumberString(a, b)) {
        return $ToNumber(a) >= $ToNumber(b);
    } else {
        return $ToString(a) >= $ToString(b);
    }
};
export const $Lt = (a: VmAny, b: VmAny): boolean => {
    if (overloadNumberString(a, b)) {
        return $ToNumber(a) < $ToNumber(b);
    } else {
        return $ToString(a) < $ToString(b);
    }
};
export const $Lte = (a: VmAny, b: VmAny): boolean => {
    if (overloadNumberString(a, b)) {
        return $ToNumber(a) <= $ToNumber(b);
    } else {
        return $ToString(a) <= $ToString(b);
    }
};
export const $Eq = (a: VmAny, b: VmAny): boolean => {
    $AssertInit(a);
    $AssertInit(b);
    // Number comparison is a special case to handle NaN correctly
    if (typeof a == 'number' && typeof b == 'number') return a === b;
    return isSame(a, b);
};
export const $Neq = (a: VmAny, b: VmAny): boolean => {
    return !$Eq(a, b);
};
export const $Aeq = (a: VmAny, b: VmAny): boolean => {
    if (overloadNumberString(a, b)) {
        const an = $ToNumber(a);
        const bn = $ToNumber(b);
        const EPS = 1e-15;
        if (isNaN(an) || isNaN(bn)) return false;
        // Since Inf - Inf is NaN, we must check for equality first
        if (an === bn) return true;
        const absoluteDifference = abs(an - bn);
        if (absoluteDifference < EPS) return true;
        const base = min(abs(an), abs(bn));
        return absoluteDifference < base * EPS;
    } else {
        // For strings, we use normalized case-insensitive comparison
        const as = $ToString(a);
        const bs = $ToString(b);
        if (as === bs) return true;
        const ai = as.toLowerCase();
        const bi = bs.toLowerCase();
        if (ai === bi) return true;
        const an = ai.normalize('NFC');
        const bn = bi.normalize('NFC');
        return an === bn;
    }
};
export const $Naeq = (a: VmAny, b: VmAny): boolean => {
    return !$Aeq(a, b);
};
export const $Same = (a: VmAny, b: VmAny): boolean => {
    $AssertInit(a);
    $AssertInit(b);
    return isSame(a, b);
};
export const $Nsame = (a: VmAny, b: VmAny): boolean => {
    return !$Same(a, b);
};
export const $In = (value: VmAny, iterable: VmAny): boolean => {
    $AssertInit(value);
    $AssertInit(iterable);
    if (iterable == null) return false;
    if (typeof iterable != 'object') return false;
    if (isVmArray(iterable)) {
        if (value == null) {
            // array may have empty slots
            for (const item of iterable) if (item == null) return true;
            return false;
        }
        // JS %SameValueZero is same with `isSame` in this context
        if (isVmPrimitive(value)) return iterable.includes(value);
        // value is not null here, so it's ok to skip empty slots, since `isSame(null, something)` is always false
        return iterable.some((item = null) => isSame(item, value satisfies NonNullable<VmValue>));
    }
    // iterable is a record or an extern here, value should be a string
    const key = toString(value, undefined);
    if (isVmWrapper(iterable)) return iterable.has(key);
    return hasOwnEnumerable(iterable satisfies VmRecord, key);
};
export const $Concat = (...args: readonly string[]): string => {
    return args.map((a) => toFormat(a, null)).join('');
};
export const $Pos = (a: VmAny): number => {
    return $ToNumber(a);
};
export const $Neg = (a: VmAny): number => {
    return -$ToNumber(a);
};
export const $Not = (a: VmAny): boolean => {
    return !$ToBoolean(a);
};
export const $Length = (value: VmAny): number => {
    $AssertInit(value);
    if (isVmArray(value)) return value.length;
    if (isVmRecord(value)) return keys(value).length;
    if (isVmWrapper(value)) {
        return value.keys().length;
    }
    return Number.NaN;
};
export const $Omit = (value: VmAny, omitted: ReadonlyArray<number | string>): VmRecord => {
    $AssertInit(value);
    if (value == null || !isVmRecord(value)) return {};
    const result: Record<string, VmConst> = {};
    const valueKeys = keys(value);
    const omittedSet = new Set(omitted.map($ToString));
    for (const key of valueKeys) {
        if (!omittedSet.has(key)) {
            /* c8 ignore next */
            result[key] = value[key] ?? null;
        }
    }
    return result;
};
export const $Pick = (value: VmAny, picked: ReadonlyArray<number | string>): VmRecord => {
    $AssertInit(value);
    if (value == null || !isVmRecord(value)) return {};
    const result: Record<string, VmConst> = {};
    for (const key of picked) {
        const k = $ToString(key);
        if (hasOwnEnumerable(value, k)) {
            result[k] = value[k] ?? null;
        }
    }
    return result;
};

const sliceCore = (value: VmArray, start: number, end: number, exclusive: boolean): VmArray => {
    const { length } = value;

    if (isNaN(start)) start = 0;
    else if (start < 0) start = length + start;
    if (isNaN(end)) end = exclusive ? length : length - 1;
    else if (end < 0) end = length + end;

    start = ceil(start);
    if (exclusive || !isSafeInteger(end)) {
        end = ceil(end);
    } else {
        end = end + 1;
    }
    return slice.call(value, start, end) as VmArray;
};
export const $Slice = (value: VmAny, start: VmAny, end: VmAny): VmArray => {
    $AssertInit(value);
    if (!isVmArray(value)) throw new VmError(`Expected array, got ${$Type(value)}`, []);
    const s = start != null ? $ToNumber(start) : 0;
    const e = end != null ? $ToNumber(end) : value.length - 1;
    return sliceCore(value, s, e, false);
};
export const $SliceExclusive = (value: VmAny, start: VmAny, end: VmAny): VmArray => {
    $AssertInit(value);
    if (!isVmArray(value)) throw new VmError(`Expected array, got ${$Type(value)}`, []);
    const s = start != null ? $ToNumber(start) : 0;
    const e = end != null ? $ToNumber(end) : value.length;
    return sliceCore(value, s, e, true);
};
export const $AssertInit: (value: VmAny) => asserts value is VmValue = (value) => {
    if (value === undefined) throw new VmError(`Uninitialized value`, null);
};
export const $Call = (func: VmValue, args: readonly VmAny[]): VmValue => {
    for (const a of args) {
        $AssertInit(a);
    }
    if (isVmExtern(func)) {
        return func.call(args as readonly VmValue[]) ?? null;
    }
    if (isVmFunction(func)) {
        return func(...(args as readonly VmValue[])) ?? null;
    }
    throw new VmError(`Value is not callable: ${display(func)}`, null);
};
export const $Type = (value: VmAny): TypeName => {
    // 允许未初始化值通过
    if (value == null) return 'nil';
    if (isVmWrapper(value)) return value.type;
    if (isVmArray(value)) return 'array';
    if (typeof value == 'object') return 'record';
    return typeof value as TypeName;
};
export const $ToBoolean = (value: VmAny): boolean => {
    if (typeof value == 'boolean') return value;
    $AssertInit(value);
    return toBoolean(value, undefined);
};
export const $ToString = (value: VmAny): string => {
    if (typeof value == 'string') return value;
    $AssertInit(value);
    return toString(value, undefined);
};
export const $ToNumber = (value: VmAny): number => {
    if (typeof value == 'number') return value;
    $AssertInit(value);
    return toNumber(value, undefined);
};
export const $IsBoolean = (value: VmAny): value is boolean => {
    $AssertInit(value);
    return typeof value == 'boolean';
};
export const $IsNumber = (value: VmAny): value is number => {
    $AssertInit(value);
    return typeof value == 'number';
};
export const $IsString = (value: VmAny): value is string => {
    $AssertInit(value);
    return typeof value == 'string';
};
export const $IsRecord = (value: VmAny): value is VmRecord => {
    $AssertInit(value);
    return isVmRecord(value);
};
export const $IsArray = (value: VmAny): value is VmArray => {
    $AssertInit(value);
    return isVmArray(value);
};
export const $AssertNonNil = (value: VmAny): asserts value is NonNullable<VmValue> => {
    $AssertInit(value);
    if (value !== null) return;
    throw new VmError('Expected non-nil value', null);
};
export const $Has = (obj: VmAny, key: VmAny): boolean => {
    $AssertInit(obj);
    const pk = $ToString(key);
    if (obj == null || typeof obj != 'object') return false;
    if (isVmWrapper(obj)) return obj.has(pk);
    return hasOwnEnumerable(obj, pk);
};
export const $Get = (obj: VmAny, key: VmAny): VmValue => {
    $AssertInit(obj);
    if (isVmArray(obj)) {
        const index = $ToNumber(key);
        if (isNaN(index)) return null;
        return (at.call(obj, trunc(index)) as VmConst | undefined) ?? null;
    }
    const pk = $ToString(key);
    if (obj == null || typeof obj != 'object') return null;
    if (isVmWrapper(obj)) return obj.get(pk) ?? null;
    if (!hasOwnEnumerable(obj, pk)) return null;
    return (obj as Record<string, VmImmutable>)[pk] ?? null;
};
export const $Set = (obj: VmAny, key: VmAny, value: VmAny): void => {
    $AssertInit(obj);
    $AssertInit(value);
    const pk = $ToString(key);
    if (obj == null) return;
    if (!isVmExtern(obj)) throw new VmError(`Expected extern, got ${$Type(obj)}`, undefined);
    obj.set(pk, value);
};
export const $Iterable = (value: VmAny): Iterable<VmValue | undefined> => {
    $AssertInit(value);
    if (isVmWrapper(value)) return value.keys();
    if (isVmArray(value)) return value;
    if (value != null && typeof value == 'object') return keys(value);
    throw new VmError(`Value is not iterable`, isVmFunction(value) ? [] : [value]);
};

export const $RecordSpread = (record: VmAny): VmRecord | null => {
    $AssertInit(record);
    if (record == null || isVmRecord(record)) return record;
    if (isVmArray(record)) {
        const result: Record<string, VmConst> = {};
        const len = record.length;
        for (let i = 0; i < len; i++) {
            const item = record[i];
            result[i] = item ?? null;
        }
        return result;
    }
    if (isVmExtern(record)) {
        const result: Record<string, VmConst> = create(null);
        for (const key of record.keys()) {
            const value = record.get(key) ?? null;
            // 当前只有 Primitive 不会进行二次包装
            if (isVmPrimitive(value)) {
                result[key] = value;
            }
        }
        return result;
    }
    throw new VmError(`Expected record, array, extern or nil, got ${$Type(record)}`, null);
};

export const $ArraySpread = (array: VmAny): Iterable<VmConst | undefined> => {
    $AssertInit(array);
    if (array == null) return [];
    if (isVmArray(array)) return array;
    if (isVmExtern(array) && typeof (array.value as Iterable<unknown>)[Symbol.iterator] == 'function') {
        const result: VmConst[] = [];
        for (const item of array.value as Iterable<unknown>) {
            // 当前只有 Primitive 不会进行二次包装
            if (isVmPrimitive(item)) {
                result.push(item);
            } else {
                result.push(null);
            }
        }
        return result;
    }
    throw new VmError(`Expected array, iterable extern or nil, got ${$Type(array)}`, []);
};

export const $Format = (value: VmAny, format: VmAny): string => {
    $AssertInit(value);
    const f = format == null ? '' : $ToString(format);
    return toFormat(value, f);
};

File: crates/python/mirascript/vm/operations.py
import math
import unicodedata
from enum import Enum
from typing_extensions import Callable, Mapping, Any

from ..helpers.convert.to_boolean import toBoolean
from ..helpers.convert.to_format import toFormat
from ..helpers.convert.to_number import toNumber
from ..helpers.checker import is_safe_integer
from ..helpers.convert.to_string import toString
from ..helpers.constants import kVmScript
from ..helpers.types import (
    is_vm_primitive,
    is_vm_record,
    is_vm_extern,
    is_vm_array,
    is_vm_module,
    is_vm_wrapper,
)
from .error import VmError
from .types.module import VmModule
from .types.function import vm_function
from .types.types import Uninitialized, VmValue


class LoopControl(Enum):
    Continue = object()
    Break = object()


## 标记当前值未返回的值
LoopContinue = LoopControl.Continue
## 标记当前值为Break
LoopBreak = LoopControl.Break


def IsNumber(a):
    if isinstance(a, bool):
        return False
    return isinstance(a, (int, float))


def isSame(a, b) -> bool:
    if IsNumber(a) and IsNumber(b):
        return a == b or (math.isnan(a) and math.isnan(b))
    if a == b and type(a) is type(b):
        return True
    if is_vm_wrapper(a):
        return a.same(b)
    if is_vm_wrapper(b):
        return b.same(a)
    return False


def AssertInit(val):
    if val is Uninitialized:
        raise VmError("Uninitialized value`", None)


def ToBoolean(value):
    AssertInit(value)
    return toBoolean(value)


def ToNumber(value):
    """
    将输入值转换为数字

    参数:
        value: 要转换的值
        default: 转换失败时返回的默认值
        prefer_float: 是否优先转换为浮点数

    返回:
        转换后的数字，如果无法转换则返回默认值
    """
    AssertInit(value)
    return toNumber(value)


def ToString(val) -> str:
    AssertInit(val)
    if val is None:
        return ""
    return toString(val)


def _overloadNumberString(a, b):
    if IsNumber(a) or IsNumber(b):
        return True
    if isinstance(a, str) or isinstance(b, str):
        return False
    return True


def Add(a, b) -> float:
    return ToNumber(a) + ToNumber(b)


def Sub(a, b) -> float:
    return ToNumber(a) - ToNumber(b)


def Mul(a, b) -> float:
    return ToNumber(a) * ToNumber(b)


def Div(a, b) -> float:
    a = ToNumber(a)
    b = ToNumber(b)
    if b != 0:
        return a / b

    if a == 0:
        return math.nan
    return math.copysign(1, a) * math.copysign(1, b) * math.inf


def Mod(a, b) -> float:
    x = ToNumber(a)
    y = ToNumber(b)
    """
    IEEE 754 标准的取余运算 (remainder operation)
    对应 ECMAScript 的 % 运算符
    
    规范定义：
    1. If either operand is NaN, the result is NaN.
    2. If x is ±Infinity, the result is NaN.
    3. If y is ±Infinity and x is finite, the result is x.
    4. If y is ±0, the result is NaN.
    5. If x is ±0 and y is nonzero and finite, the result is x.
    6. Otherwise, the result has the same sign as x and magnitude:
       abs(x) - floor(abs(x) / abs(y)) × abs(y)
    """
    # 1. If either operand is NaN, return NaN
    if math.isnan(x) or math.isnan(y):
        return math.nan

    # 2. If x is ±Infinity, return NaN
    if math.isinf(x):
        return math.nan

    # 3. If y is ±Infinity and x is finite, return x
    if math.isinf(y):
        return x

    # 4. If y is ±0, return NaN
    if y == 0:
        return math.nan

    # 5. If x is ±0 and y is nonzero and finite, return x (保留符号)
    if x == 0:
        return x  # 保留 +0 或 -0 的符号

    # 6. 计算取余：结果符号与 x 相同
    # Python 的 % 运算符结果符号与除数相同，不符合 IEEE 754
    # 需要使用 math.fmod，它的结果符号与被除数相同
    return math.fmod(x, y)


def Pow(a, b) -> float:
    try:
        return math.pow(ToNumber(a), ToNumber(b))
    except ValueError:
        return math.nan


def And(a, b) -> bool:
    return ToBoolean(a) and ToBoolean(b)


def Or(a, b) -> bool:
    return ToBoolean(a) or ToBoolean(b)


def Gt(a, b) -> bool:
    if _overloadNumberString(a, b):
        return ToNumber(a) > ToNumber(b)
    else:
        return str(a) > str(b)


def Gte(a, b) -> bool:
    if _overloadNumberString(a, b):
        return ToNumber(a) >= ToNumber(b)
    else:
        return str(a) >= str(b)


def Lt(a, b) -> bool:
    if _overloadNumberString(a, b):
        return ToNumber(a) < ToNumber(b)
    else:
        return str(a) < str(b)


def Lte(a, b) -> bool:
    if _overloadNumberString(a, b):
        return ToNumber(a) <= ToNumber(b)
    else:
        return str(a) <= str(b)


def Eq(a, b) -> bool:
    AssertInit(a)
    AssertInit(b)

    if isinstance(a, (int, float)) and isinstance(b, (int, float)):
        return a == b

    return isSame(a, b)


def Neq(a, b) -> bool:
    return not Eq(a, b)


## 高级相等比较
def Aeq(a, b) -> bool:
    if _overloadNumberString(a, b):
        an = ToNumber(a)
        bn = ToNumber(b)
        EPS = 1e-15
        if math.isnan(an) or math.isnan(bn):
            return False
        if an == bn:
            return True
        absoluteDifference = abs(an - bn)
        if absoluteDifference < EPS:
            return True
        base = min(abs(an), abs(bn))
        return absoluteDifference < base * EPS
    else:
        as_str = ToString(a)
        bs_str = ToString(b)

        if as_str == bs_str:
            return True

        ai = unicodedata.normalize("NFC", as_str.lower())
        bi = unicodedata.normalize("NFC", bs_str.lower())
        return ai == bi


def Naeq(a, b) -> bool:
    return not Aeq(a, b)


def Same(a, b) -> bool:
    AssertInit(a)
    AssertInit(b)
    return isSame(a, b)


def Nsame(a, b) -> bool:
    return not Same(a, b)


def In(a, b) -> bool:
    AssertInit(a)
    AssertInit(b)
    if b is None:
        return False
    if is_vm_array(b):
        if a is None:
            return a in b

        if is_vm_primitive(a):
            return a in b

        for item in b:
            if isSame(a, item):
                return True
        return False
    pk = ToString(a)
    if isinstance(b, dict):
        return pk in b
    if is_vm_wrapper(b):
        return b.has(pk)

    return False


def Concat(*args) -> str:
    result = ""
    for a in args:
        AssertInit(a)
        result += toFormat(a, None)
    return result


def Pos(a) -> float:
    return ToNumber(a)


def Neg(a) -> float:
    return -ToNumber(a)


def Not(a) -> bool:
    return not ToBoolean(a)


def Length(a) -> float:
    AssertInit(a)
    if isinstance(a, (str, list, dict)):
        return float(len(a))
    raise TypeError(f"`Expected array, string or record, got {Type(a)}")


def Omit(a, b):
    AssertInit(a)
    if a is None or not is_vm_record(a):
        return {}
    result = {}

    valueKeys = a.keys()
    omittedSet = set([ToString(x) for x in b])
    for key in valueKeys:
        if key not in omittedSet:
            result[key] = a[key]
    return result


def Pick(a, b):
    AssertInit(a)
    if a is None or not is_vm_record(a):
        return {}
    result = {}
    for key in b:
        k = ToString(key)
        if k in a:
            result[k] = a[k]

    return result


def _slice(a, start, end, exclusive):
    length = len(a)

    if math.isnan(start):
        start = 0
    elif start < 0:
        start = length + start
    if math.isnan(end):
        end = length
    elif end < 0:
        end = length + end
    if not math.isinf(start):
        start = math.ceil(start)
    else:
        if start > 0:
            start = length
        else:
            start = 0

    if math.isinf(end):
        if end > 0:
            end = length
        else:
            end = 0

    elif exclusive or not is_safe_integer(end):
        end = math.ceil(end)
    else:
        end = int(end + 1)
    return a[start:end]


def Slice(a, start, end):
    AssertInit(a)
    if not is_vm_array(a):
        raise VmError(f"`Expected array, got {Type(a)}", [])
    s = ToNumber(start) if start is not None else 0
    e = ToNumber(end) if end is not None else len(a) - 1
    return _slice(a, s, e, False)


def SliceExclusive(a, start, end):
    AssertInit(a)
    if not is_vm_array(a):
        raise VmError(f"`Expected array, got {Type(a)}", [])
    s = ToNumber(start) if start is not None else 0
    e = ToNumber(end) if end is not None else len(a)
    return _slice(a, s, e, True)


def Call(func, *args):
    for a in args:
        AssertInit(a)
    # if isinstance(func, VmExtern) and hasattr(func, "callable"):
    #     return func.call(args)
    if callable(func):
        if len(args) == 0:
            return func()
        return func(*args)
    raise VmError(f"{type(func)}, {func} object is not callable", None)


def Type(val):
    if val is Uninitialized or val is None:
        return "nil"
    if isinstance(val, bool):
        return "boolean"
    if isinstance(val, (int, float)):
        return "number"
    if isinstance(val, str):
        return "string"
    if is_vm_array(val):
        return "array"
    if is_vm_record(val):
        return "record"
    if is_vm_extern(val):
        return "extern"
    if is_vm_module(val):
        return "module"
    if callable(val):
        return "function"
    return type(val).__name__


def IsBoolean(val):
    AssertInit(val)
    return isinstance(val, bool)


def IsString(val):
    AssertInit(val)
    return isinstance(val, str)


def IsRecord(val):
    AssertInit(val)
    return is_vm_record(val)


def IsArray(val):
    AssertInit(val)
    return is_vm_array(val)


def AssertNonNil(val):
    AssertInit(val)
    if val is not None:
        return
    raise VmError("Expected non-nil value", None)


def Has(obj, key):
    AssertInit(obj)
    pk = ToString(key)
    if obj is None:
        return False

    if is_vm_wrapper(obj):
        return obj.has(pk)
    if isinstance(obj, dict):
        return pk in obj
    if isinstance(obj, (list, tuple)):
        try:
            idx = int(float(key))
            return 0 <= idx < len(obj)
        except:
            return False
    if hasattr(obj, pk):
        return True
    return False


def Get(obj, key):
    AssertInit(obj)
    pk = ToString(key)
    if obj is None:
        return None

    if is_vm_wrapper(obj):
        return obj.get(pk)
    if isinstance(obj, dict):
        return obj.get(pk, None)
    if isinstance(obj, (list, tuple)):
        try:
            idx = int(float(key))
            return obj[idx]
        except:
            return None
    if hasattr(obj, pk):
        return getattr(obj, pk)
    return None


def Set(obj, key, val):
    AssertInit(obj)
    AssertInit(val)
    pk = ToString(key)
    if obj is None:
        return
    if not is_vm_extern(obj):
        raise VmError(f"`Expected extern object, got {Type(obj)}", None)

    obj.set(pk, val)


def Iterable(val):
    AssertInit(val)
    if is_vm_wrapper(val):
        if hasattr(val, "keys"):
            return val.keys()
    if isinstance(val, (list, tuple)):
        return val
    if isinstance(val, dict):
        return val.keys()
    raise VmError(f"`Value is not iterable {Type(val)}", None)


def RecordSpread(val):
    AssertInit(val)
    if val is None:
        return {}
    if is_vm_record(val):
        return val
    if is_vm_array(val):
        return {ToString(i): val[i] for i in range(len(val))}
    if is_vm_extern(val):
        return {}

    raise VmError(f"`Expected record, extern or nil, got {Type(val)}", None)


def ArraySpread(val):
    AssertInit(val)
    if val is None:
        return []
    if is_vm_array(val):
        return val

    raise VmError(f"`Expected array, iterable extern or nil, got {Type(val)}", None)


def Format(val, fmt=None):
    AssertInit(val)

    return toFormat(val, fmt)


def Range(start, end):
    s = ToNumber(start)
    e = ToNumber(end)
    if math.isnan(s) or math.isnan(e):
        return []
    result = []
    while s <= e:
        result.append(s)
        s += 1
    return result


def RangeExclusive(start, end):
    pass
    s = ToNumber(start)
    e = ToNumber(end)
    if math.isnan(s) or math.isnan(e):
        return []
    result = []
    while s < e:
        result.append(s)
        s += 1
    return result


def Fn(name):

    def decorator(func):

        from .helpers import CpEnter, CpExit

        fn = func

        def fn_wrapper(*args, **kwargs):
            try:
                CpEnter()
                return fn(*args, **kwargs)
            finally:
                CpExit()

        return vm_function(name)(fn_wrapper)

    return decorator


def Closure(func):

    from .helpers import CpEnter, CpExit

    def closure_wrapper():
        try:
            CpEnter()
            return func()
        finally:
            CpExit()

    return closure_wrapper


def Script(func):

    from mirascript.vm.helpers import CpEnter, CpExit

    def script_wrapper(*args, **kwargs):
        try:
            CpEnter()
            return func(*args, **kwargs)
        finally:
            CpExit()

    setattr(script_wrapper, kVmScript, True)
    return script_wrapper


_PUB_ATTR = "__mirascript.mod.pub__"


def Module(name: str):

    from .helpers import Upvalue

    def decorator(kls: type):

        pub: "dict[str, Callable[[], Any]]" = {}
        for attr_name in dir(kls):
            attr = getattr(kls, attr_name)
            pub_name = getattr(attr, _PUB_ATTR, None)
            if pub_name is not None:
                assert callable(attr), f"Public attribute {attr_name} must be callable"
                pub[pub_name] = attr

        class Mod(Mapping[str, VmValue]):

            def __getitem__(self, key: str) -> VmValue:
                getter = pub[key]
                return Upvalue(getter())

            def __iter__(self):
                return iter(pub.keys())

            def __len__(self):
                return len(pub)

        return VmModule(name, Mod())

    return decorator


def Pub(name: str):

    def decorator(method: Callable):
        setattr(method, _PUB_ATTR, name)
        return method

    return decorator
